### ISSUE TITLE
CustomValidatorAttribute for Razor Pages PageModel

### ISSUE THREAD

**Issue by DavidJohnWilliams** [FluentValidation/FluentValidation#1541]
Hi,

Reading the docs about page models https://docs.fluentvalidation.net/en/latest/aspnet.html#use-with-page-models. There are three things that are currently "not possible".

I would like to try and resolve these issues, however I am pretty much an open source noob (note: **not** .net noob).

As a first step I have created a proof of concept which allows `CustomValidatorAttribute` on a `PageModel`. I have added a test and it seems to work just fine and all tests pass.

As per the guidelines I am opening this issue to discuss prior to creating a pull request. However the code is ready to go.

I do have a couple of questions:

1. I have the extension Code Cleanup on Save installed and when saving certain files it was applying **my** editor config in some cases. This caused a number of formatting changes that are unrelated to my code changes. To resolve this I have made a few changes to the editor config. Which I can either submit as a separate commit or PR. However... there are also some places where these cannot be resolved with editor config because the formatting for a file is inconsistent with the rest of the codebase. e.g. remove and sort using statements or explicit type vs `var`. So the question is should I commit those formatting changes in a separate commit or even separate PR or as they are really fairly minor or can they be included with this change?

2. In order to add this feature ASP.NET Core 2.X seems to require the package `Microsoft.AspNetCore.Mvc.RazorPages` whereas ASP.NET Core 3.X includes it in the framework. I noticed that there is only one other package at present which is `Microsoft.AspNetCore.Mvc.ViewFeatures` so I am not sure if this was the right thing?

Look forward to hearing from you!

Best regards

Dave

**Comment by JeremySkinner** [comment]
Hi Dave, thanks so much for being willing to contribute! If you could open a PR that would be great. To answer your questions:

> To resolve this I have made a few changes to the editor config. Which I can either submit as a separate commit or PR. 

If you could submit that as a separate PR that'd be great, it's much easier to review things when there aren't unrelated changes in the same PR. 

> So the question is should I commit those formatting changes in a separate commit or even separate PR or as they are really fairly minor or can they be included with this change?

Are they in the same files that you've changed to enable this new functionality? If so (and if they're minor) then let's keep them in this PR for now and I'll decide when I review.

> In order to add this feature ASP.NET Core 2.X seems to require the package Microsoft.AspNetCore.Mvc.RazorPages whereas ASP.NET Core 3.X includes it in the framework. I noticed that there is only one other package at present which is Microsoft.AspNetCore.Mvc.ViewFeatures so I am not sure if this was the right thing?

We decided to explicitly remove the dependency on razor when targetting netcore 2.x, as it brings in several large packages which aren't typically needed (eg, in a web api project). Is it possible to achieve what you're doing without referencing `Microsoft.AspNetCore.Mvc.RazorPages`? If not then it may be better to only enable this new feature when targetting netcore3.1.

Once you've submitted the PR I'll give it a review and we can go from there :) 

Thanks!

**Comment by JeremySkinner** [comment]
Also, if possible it'd be great if you could base these changes on top of the `net5` branch (rather than master) so we can ensure that they work with the upcoming .net 5 release too. Thanks!

**Comment by DavidJohnWilliams** [comment]
Thanks for the quick response. I will get the PR's submitted ASAP. Once I figure out how you do that...

With regards to the use of `Microsoft.AspNetCore.Mvc.RazorPages`, no I don't think it can be done without it. I did originally wrap all of my code in `#if NETCOREAPP3_0 || NETCOREAPP3_1` and that all worked. But I wasn't sure about the tests. I could exclude the test in 2.X?

However... is there any actually any use case for `FluentValidation.AspNetCore` where you wouldn't have installed `Microsoft.AspNetCore.Mvc` which brings in `Microsoft.AspNetCore.Mvc.RazorPages` anyway?


With regards to your most comment about branches... do I need to commit to the net5 branch or a branch of my own taken from that one? I'm not that familiar with how branching in git works (we use TFVC but are currently moving over).

**Comment by JeremySkinner** [comment]
> I did originally wrap all of my code in #if NETCOREAPP3_0 || NETCOREAPP3_1 and that all worked. But I wasn't sure about the tests. I could exclude the test in 2.X?

Yes, let's exclude the test in 2.x. You'll want to use `#if NETCOREAPP3_1 || NET5_0` for (3.0 is out of support, so no need to have an explicit case for this)

> However... is there any actually any use case for FluentValidation.AspNetCore where you wouldn't have installed Microsoft.AspNetCore.Mvc

Yes, in netcore2.x you may have scenarios where you're building web apis only, so you don't want to bring along and deploy all the razor packages when your application doesn't have any views. With 3.x this is no longer an issue as they all come as part of the runtime, but with 2.x you had to deploy them all manually. 

> do I need to commit to the net5 branch or a branch of my own taken from that one?

Either would work, but typically you'd create your own branch based on that one to hold your changes, and then submit a pull request for me to merge it into net5 at my end:

```
# First checkout the net5 branch
git checkout net5

# Then create your own branch based on net5 (using an example branch name of "feature/razor-pages-customize-validator")
git checkout -b feature/razor-pages-customize-validator

# Make your changes
git add -A
git commit

# Push to your fork of the repository using the branch name created earlier
git push origin feature/razor-pages-customize-validator

# Then you can create a PR via the github ui
```

Give me a shout if you get stuck

**Comment by DavidJohnWilliams** [comment]
@JeremySkinner FYI I had to download VS 2019 preview to use .Net 5.0. That meant the aforementioned extension was no longer installed. If I work on this repo in future (which I hope to do) then I will submit the changes for editor config at a later date. It would probably be worth some discussion and then applying the formatting changes to the whole solution rather than piecemeal.

**Comment by JeremySkinner** [comment]
The .net5 branch has been merged into master, and this will be included in the 9.3 release.
